src/audio/pipeline/pipeline-stream.c

```diff
@@ -299,17 +301,129 @@ static void pipeline_trigger_xrun(struct pipeline *p, struct comp_dev **host)
 	} while (true);
 }
 
+#if CONFIG_KCPS_DYNAMIC_CLOCK_CONTROL
+static int add_pipeline_cps_consumption(struct comp_dev *current,
+					struct comp_buffer *calling_buf,
+					struct pipeline_walk_context *ctx, int dir)
+{
+	struct pipeline_data *ppl_data = ctx->comp_data;
+	struct ipc4_base_module_cfg *cd;
+	int kcps;
+
+	pipe_dbg(ppl_data->p, "add_pipeline_cps_consumption(), current->comp.id = %u, dir = %u",
+		 dev_comp_id(current), dir);
+
+	if (!comp_is_single_pipeline(current, ppl_data->start)) {
+		pipe_dbg(ppl_data->p, "add_pipeline_cps_consumption(), current is from another pipeline");
+		return 0;
+	}
+
+	/* modules created through module adapter have different priv_data */
+	if (current->drv->type != SOF_COMP_MODULE_ADAPTER) {
+		cd = comp_get_drvdata(current);
+	} else {
+		struct processing_module *mod = comp_get_drvdata(current);
+		struct module_data *md = &mod->priv;
+
+		cd = &md->cfg.base_cfg;
+	}
+
+	if (cd->cpc == 0) {
+		/* Use maximum clock budget, assume 1ms chunk size */
+		ppl_data->kcps = CLK_MAX_CPU_HZ / 1000;
+		tr_warn(pipe,
+			"0 CPS requested for module: %#x, core: %d using safe max KCPS: %u",
+			current->ipc_config.id, ppl_data->p->core, ppl_data->kcps);
+
+		return PPL_STATUS_PATH_STOP;
+	} else {
+		kcps = cd->cpc * 1000 / ppl_data->p->period;
+		tr_info(pipe, "Registering module: %#x KCPS consumption: %d, core: %d",
+			current->ipc_config.id, kcps, ppl_data->p->core);
+
+		ppl_data->kcps += kcps;
+	}
+
+	return pipeline_for_each_comp(current, ctx, dir);
+}
+
+static int remove_pipeline_cps_consumption(struct comp_dev *current,
+					   struct comp_buffer *calling_buf,
+					   struct pipeline_walk_context *ctx, int dir)
+{
+	struct pipeline_data *ppl_data = ctx->comp_data;
+	struct ipc4_base_module_cfg *cd;
+	int kcps;
+
+	pipe_dbg(ppl_data->p, "remove_pipeline_cps_consumption(), current->comp.id = %u, dir = %u",
+		 dev_comp_id(current), dir);
+
+	if (!comp_is_single_pipeline(current, ppl_data->start)) {
+		pipe_dbg(ppl_data->p, "remove_pipeline_cps_consumption(), current is from another pipeline");
+		return 0;
+	}
+
+	/* modules created through module adapter have different priv_data */
+	if (current->drv->type != SOF_COMP_MODULE_ADAPTER) {
+		cd = comp_get_drvdata(current);
+	} else {
+		struct processing_module *mod = comp_get_drvdata(current);
+		struct module_data *md = &mod->priv;
+
+		cd = &md->cfg.base_cfg;
+	}
+
+	kcps = cd->cpc * 1000 / ppl_data->p->period;
+	tr_info(pipe, "Unregistering module: %#x KCPS consumption: %d, core: %d",
+		current->ipc_config.id, kcps, ppl_data->p->core);
```

Comment:
maybe consider making this and a similar print in `add_pipeline_cps_consumption()` above a `tr_dbg()`